app/models/papi_client.rb

```diff
@@ -0,0 +1,307 @@
+  def create_virtual_machine_object(machine_type: 'n1-standard-1', preemptible: false)
+    Google::Apis::GenomicsV2alpha1::VirtualMachine.new(
+        machine_type: machine_type,
+        preemptible: preemptible,
+        service_account: Google::Apis::GenomicsV2alpha1::ServiceAccount.new(email: self.issuer, scopes: GOOGLE_SCOPES)
+    )
+  end
+
+  # Determine command line to pass to ingest based off of file & action requested
+  #
+  # * *params*
+  #   - +study_file+ (StudyFile) => StudyFile to be ingested
+  #   - +action+ (String/Symbol) => Action to perform on ingest
+  #
+  # * *return*
+  #   - (Array) Command Line, in Docker "exec" format
+  #
+  # * *raises*
+  #   - (ArgumentError) => The requested StudyFile and action do not correspond with each other, or cannot be run yet
+  def get_command_line(study_file:, action:)
+    validate_action_by_file(action, study_file)
+    study = study_file.study
+    command_line = "python ingest_pipeline.py --study-accession #{study.accession} --file-id #{study_file.id} #{action}"
+    case action.to_s
+    when 'ingest_expression'
+      if study_file.file_type == 'Expression Matrix'
+        command_line += " --matrix-file #{study_file.gs_url} --matrix-file-type dense"
+      elsif study_file.file_type === 'MM Coordinate Matrix'
+        bundled_files = study_file.bundled_files
+        genes_file = bundled_files.detect {|f| f.file_type == '10X Genes File'}
+        barcodes_file = bundled_files.detect {|f| f.file_type == '10X Barcodes File'}
+        command_line += " --matrix-file #{study_file.gs_url} --matrix-file-type mtx" \
+                      " --gene-file #{genes_file.gs_url} --barcode-file #{barcodes_file.gs_url}"
+      end
+    when 'ingest_cell_metadata'
+      command_line += " --cell-metadata-file #{study_file.gs_url} --ingest-cell-metadata"
+    when 'ingest_cluster'
+      command_line += " --cluster-file #{study_file.gs_url} --ingest-cluster"
+    when 'ingest_subsample'
+      metadata_file = study.metadata_file
+      command_line += " --cluster-file #{study_file.gs_url} --cell-metadata-file #{metadata_file.gs_url} --subsample"
+    end
+
+
+    # add optional command line arguments based on file type
+    optional_args = self.get_command_line_options(study_file)
+    # return an array of tokens (Docker expects exec form, which runs without a shell, so cannot be a single command)
+    exec_form = command_line.split + optional_args
+    exec_form
+  end
+
+  # Assemble any optional command line options for ingest by file type
+  #
+  # * *params*
+  #   - +study_file+ (StudyFile) => File to be ingested
+  #
+  # * *returns*
+  #   (Array) => Array representation of optional arguments (Docker exec form), based on file type
+  def get_command_line_options(study_file)
+    opts = []
+    case study_file.file_type
+    when /Matrix/
+      taxon = study_file.taxon
+      opts += ["--taxon-name", "#{taxon.scientific_name}", "--taxon-common-name", "#{taxon.common_name}",
```

Comment:
Could you add a comment in the code explaining what is passed for these three fields when their value is empty / null?  (I forget if it's an empty string; a code comment would ease recall.)

Reply:
Can do.